feat: track response time of `invoke` and `sendSync` methods on `ipcRenderer`

hitarth-gg opened the pull request:
#### Description of Change
- Allows Devtron to track the time (in milliseconds) it takes for `ipcRenderer.invoke` and `ipcRenderer.sendSync` to receive responses back from the main process.
- Patches `.on`, `.off`, `.once`, `.addListener`, `.removeListener`, `.removeAllListeners`, `.handle`, `.handleOnce`, and `.removeHandler` methods of `ipcMain` to allow Devtron to track response times and detect when handlers or listeners are **removed** from `ipcMain`.

#### Tracking `.sendSync` and `.invoke` response times:
| Request | Response |
| ------------- | ------------- |
| <img width="1142" height="579" alt="image" src="https://github.com/user-attachments/assets/9e4bb1ba-843a-479b-9433-702e2bc6424f" />  |  <img width="1142" height="579" alt="image" src="https://github.com/user-attachments/assets/d6bbbd0f-fdfe-4ac0-b2a4-f16850fc3d8a" /> |

- Clicking on the `Linked Event` field in the Details Panel will redirect the user to the linked event.
- Internally, the two events are linked using a UUID. Renderer process generates a UUID and includes it in the arguments of `.sendSync` and `.invoke` requests. This UUID is then received by the main process, where it is extracted from the arguments. As a result, both processes use the same UUID to track the events. Events with matching UUIDs are then linked together in the frontend.

#### Tracking `.once` and `.handleOnce`:
<img width="1142" height="307" alt="image" src="https://github.com/user-attachments/assets/ce95a872-35e7-4c19-bcf5-948900515f7d" />

- `ipcMain.once` is tracked as two separate events: `ipcMain.removeListener` and `ipcMain.on`. 
- `ipcMain.handleOnce` is tracked as two separate events: `ipcMain.removeHandler` and `ipcMain.handle`.

samuelmaddock commented on src/types/shared.ts:
```diff
@@ -5,7 +5,9 @@ export type Direction =
   | 'main-to-service-worker'
-  | 'renderer';
+  | 'renderer'
+  | 'renderer'
```
duplicate
```suggestion
```

samuelmaddock commented on src/lib/electron-renderer-tracker.ts:
```diff
@@ -1,6 +1,8 @@
 import { ipcRenderer } from 'electron';
 import { MSG_TYPE } from '../common/constants';
 import type { Direction, IpcEventData } from '../types/shared';
+import { performance } from 'node:perf_hooks';
```
`performance` should be available on the global object in the renderer
```suggestion
```

hitarth-gg replied:
Resolved here: [f47bca4](https://github.com/electron-userland/devtron/pull/275/commits/f47bca418841d72015498dae6350e74420c5d33a)

samuelmaddock commented on src/lib/electron-renderer-tracker.ts:
```diff
@@ -129,18 +133,55 @@ export function monitorRenderer(): void {
   };
+
+  ipcRenderer.sendSync = function (channel: string, ...args: any[]) {
+    const uuid = uuidv4(); // uuid is used to match the response with the request
```
[randomUUID](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/randomUUID) is built into browsers and node which avoids the need for the extra module.
```suggestion
    const uuid = crypto.randomUUID(); // uuid is used to match the response with the request
```

samuelmaddock commented on src/extension/react/components/Panel.tsx:
```diff
@@ -167,8 +217,17 @@ function Panel() {
-          return params.value; // or a custom string
+        cellRenderer: (params: ICellRendererParams<IpcEventDataIndexed>) => {
+          return (
+            <div className="" title={params.value}>
```
```suggestion
            <div title={params.value}>
```

samuelmaddock commented on src/index.ts:
```diff
@@ -176,10 +201,169 @@ async function startServiceWorker(ses: Electron.Session, extension: Electron.Ext
+    const cleanedListener = async (event: Electron.IpcMainInvokeEvent, ...args: any[]) => {
+      const newArgs = getArgsFromPayload(args);
+      const result = await listener(event, ...newArgs);
+      return result;
```
It could be interesting to track the `result` as well so it can be later inspected. Not blocking for this PR.

hitarth-gg replied:
Added it to the [issue](https://github.com/electron-userland/devtron/issues/272). I can probably do it in a follow up PR.

yangannyx commented on src/index.ts:
```diff
@@ -176,10 +201,169 @@ async function startServiceWorker(ses: Electron.Session, extension: Electron.Ext
+    listenerMap.get(channel)!.set(original, tracked);
+  };
+
+  const getArgsFromPayload = (payload: any[]): any[] => {
+    if (payload[0] && typeof payload[0] === 'object' && payload[0].__uuid__devtron) {
+      // If the first argument is an object with __uuid__devtron, return its args property
+      return payload[0].args || [];
+    }
+    // Otherwise, return the payload as is
+    return payload;
+  };
```
nit: it would be nice if we could extend/reuse this helper function above in `trackIpcEvent`

hitarth-gg replied:
Fixed: [18adb6d](https://github.com/electron-userland/devtron/pull/275/commits/18adb6d4979adca825202d86a1e3c6675225342e)
Lmk if it still needs changes.

yangannyx commented on src/index.ts:
```diff
@@ -176,10 +201,169 @@ async function startServiceWorker(ses: Electron.Session, extension: Electron.Ext
 }
 
+function patchIpcMain() {
+  const listenerMap = new Map<string, Map<any, any>>(); // channel -> (originalListener -> tracked/cleaned Listener)
```
Appreciate the comment here!
nit: It might be nice to create named types to make this code more self documenting:
```suggestion
// top level
type Channel = string
type Listener = any
...
  const listenerMap = new Map<Channel, Map<Listener, Listener>>(); // channel -> (originalListener -> tracked/cleaned Listener)
```

yangannyx commented on src/extension/react/components/Panel.tsx:
```diff
@@ -99,12 +130,31 @@ function Panel() {
+
         setEvents((prev) => {
-          const updated = [...prev, message.event].slice(-MAX_EVENTS_TO_DISPLAY);
+          const updated = [...prev, event].slice(-MAX_EVENTS_TO_DISPLAY);
+          // If the event with the same UUID already exists, we update it
+          if (event.uuid && uuidMapRef.current.has(event.uuid)) {
+            const index = uuidMapRef.current.get(event.uuid)!;
+            if (index < prev.length) {
+              // update the old event to include a link to the new event
+              const oldEvent = updated[index];
+              oldEvent.gotoSerialNumber = event.serialNumber;
+              updated[index] = oldEvent;
+              // add a link to the old event in the new event
+              updated[updated.length - 1] = { ...event, gotoSerialNumber: index + 1 };
+            }
+            uuidMapRef.current.delete(event.uuid);
+          } else if (event.uuid) {
+            // If a UUID is encountered for the first time, we store its index
+            uuidMapRef.current.set(event.uuid, updated.length - 1);
```
I think you'll run into a couple of issues when the events array is at `MAX_EVENTS_TO_DISPLAY`
- uuidMapRef indices might be stale and point to shifted events
- new entries to uuidMapRef will all have the same index value since updated.length is now constant

yangannyx commented on src/extension/react/components/Panel.tsx:
```diff
@@ -29,9 +30,13 @@ ModuleRegistry.registerModules([
 ]);
 import { useDevtronContext } from '../context/context';
 
+type UUID = string;
+type SerialNumber = number;
```
[non-blocking] Move these up to the top-level shared types file to tie these types with the IPCEventData shape more explicitly
```suggestion
// in src/types/shared.ts
export type UUID = string;
export type SerialNumber = number;
...
export interface IpcEventData {
    ....
    uuid?: UUID,
}
...
export interface IpcEventDataIndexed extends IpcEventData {
  serialNumber: SerialNumber;
  gotoSerialNumber?: SerialNumber; // For navigating to a specific event in the grid
}
```

hitarth-gg replied:
Fixed, this was batched in [7d5813e](https://github.com/electron-userland/devtron/pull/275/commits/7d5813e5d632d4f7883c46f06add3a3ac3f08a79).

yangannyx reviewed (approved):
PR LGTM from a functionality standpoint! Left some nonblocking suggestions around better type reusability, but don't want to bike shed on it

hitarth-gg commented on src/extension/react/components/Panel.tsx:
```diff
@@ -99,12 +142,44 @@ function Panel() {
             requestAnimationFrame(() => {
               requestAnimationFrame(() => {
-                gridRef.current?.api.ensureIndexVisible(updated.length - 1, 'bottom');
+                scrollToRow(updated.length - 1, 'bottom');
```
Fixed: [9656a9b](https://github.com/electron-userland/devtron/pull/275/commits/9656a9b0713b8913a4a98e6133da0b306dc2bf05)

hitarth-gg commented:
Rebased onto the `next` branch and resolved merge conflicts in `src/index.ts.`

hitarth-gg commented:
> Rebased onto the `next` branch and resolved merge conflicts in `src/index.ts.`
> 
> **Note**: This rebase rewrote commit hashes, so all commits show up as new. In reality, the only change since the last review is resolving merge conflicts in `src/index.ts`. No other files were modified.

Rebased again.
